exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/xsort/managed/TestSortImpl.java

```diff
@@ -466,10 +469,10 @@ public void runLargeSortTest(OperatorFixture fixture, DataGenerator dataGen,
 
   public void runJumboBatchTest(OperatorFixture fixture, int rowCount) {
     timer.reset();
-    DataGenerator dataGen = new DataGenerator(fixture, rowCount, Character.MAX_VALUE);
-    DataValidator validator = new DataValidator(rowCount, Character.MAX_VALUE);
+    DataGenerator dataGen = new DataGenerator(fixture, rowCount, ValueVector.MAX_ROW_COUNT);
```

Comment:
Well... As it turns out, `ValueVector.MAX_ROW_COUNT` is 64K, which the the maximum size an SV2 can address. (An SV2 is 16 bits wide.) `Integer.MAX_VALUE` is 2^32, which would require a 32-bit SV2, which we don't have. So, using the `Integer.MAX_VALUE` would cause the test to fail as the sorter could not sort batches larger than 64K...

Prior we used to use `Character.MAX_VALUE`, but it is not intuitively obvious that our batch size should be correlated to the size of Java's UTF-16 character encoding... And, in fact, the original bug is that they are not correlated: `Character.MAX_VALUE` is 65535, while `ValueVector.MAX_ROWS` is 65536. As a result, we were not testing the full-batch corner case.

Reply:
Correct, but will not DataGenerator/DataValidator use `Math.min(batchSize, ValueVector.MAX_ROW_COUNT);` and limit the size to the maximum possible?

Reply:
That is true. But, what advantage is there to passing an arbitrarily large number that doesn't represent a valid Drill batch size? The reader would see it as an error until they go and look at the code to see that the potential error is not really an error.

The question is: is this a bug that needs to be fixed, or just a coding preference that we can let go for now?

Reply:
It is more code maintenance question. Changing maximum value in DataGenerator/DataValidator requires similar change here.

Reply:
Thanks Vlad. I double-checked everything.

`testLargeBatch()` passes to `runJumboBatchTest()` the number of records per batch. It passes the maximum, which is `ValueVector.MAX_ROW_COUNT`:

```
      runJumboBatchTest(fixture, ValueVector.MAX_ROW_COUNT);
```

Then, `runJumboBatchTest()` sets up a test of *n* rows with *m* rows per batch. (This allows testing multiple batches with a small number of batches each.) Here, we want one batch with the maximum number of rows, so we pass:

```
    DataGenerator dataGen = new DataGenerator(fixture, rowCount, ValueVector.MAX_ROW_COUNT);
    DataValidator validator = new DataValidator(rowCount, ValueVector.MAX_ROW_COUNT);
```

That is, a row count of 64K (passed in) in batches of 64K. The result is a single batch of the maximum possible row count.

Next, `DataGenerator` records these two numbers. It was a bit forgiving by adjusting the maximum number to the allowed maximum:

```
      this.batchSize = Math.min(batchSize, ValueVector.MAX_ROW_COUNT);
```

I see that this causes confusion, and is probably a bit too lenient for a test. So, I've enforced the limit instead:

```
      Preconditions.checkArgument(batchSize > 0 && batchSize <= ValueVector.MAX_ROW_COUNT);
      this.batchSize = batchSize;
```

I made the same change to the `DataValidator` class.

We also discussed the SV4 class. What may not be entirely obvious there is the length of an SV4 can often exceed the batch size limit. We have two indirection vectors (AKA "selection vectors"). The SV2 is an indirection on top of a single batch (and so has a length limited to `ValueVector.MAX_RECORD_COUNT`.) But, the SV4 "glues together" multiple batches, and so will have a length limited only by the maximum vector length. (Hard coded at 2 GB, but ideally limited to 16 GB.)

Further, as your work showed, we probably don't want a huge SV4 for other reasons. A primary use of at the SV4 is to merge a collection if batches to produce a set of merged output batches. You showed that merge performance suffers above a certain input width, and that width is far below the theoretical maximum.

So, I *think* I've got everything covered. If I'm overlooking anything, please do point it out so I can fix it.